frame/dapps-staking/src/lib.rs

```diff
@@ -97,6 +130,128 @@ impl Default for Version {
+/// Contains information about how much was staked in each era
+#[derive(Encode, Decode, Clone, Default, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct StakerInfo<Balance: AtLeast32BitUnsigned + Copy> {
+    // Size of this list would be limited by a configurable constant TODO
+    stakes: Vec<EraStake<Balance>>,
+}
+
+impl<Balance: AtLeast32BitUnsigned + Copy> StakerInfo<Balance> {
+    /// true if no stakes exist, false otherwise
+    fn is_empty(&self) -> bool {
+        self.stakes.is_empty()
+    }
+
+    /// number of `EraStake` chunks
+    fn len(&self) -> u32 {
+        self.stakes.len() as u32
+    }
+
+    /// Stakes some value in the specified era.
+    /// User should ensure that given era is either equal or greater than the
+    /// latest available era in the staking info.
+    fn stake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if self.stakes.is_empty() {
+            self.stakes.push(EraStake::new(value, current_era))
+        } else {
+            let era_stake = self.stakes.last().unwrap(); // exists if vec not empty
+            if era_stake.era > current_era {
+                return Err("Unexpected era".into());
+            }
+
+            let new_stake_value = era_stake.staked.saturating_add(value);
+
+            if current_era == era_stake.era {
+                *self.stakes.last_mut().unwrap() = EraStake::new(new_stake_value, current_era)
+            } else {
+                self.stakes
+                    .push(EraStake::new(new_stake_value, current_era))
+            }
+        }
+        Ok(())
+    }
+
+    /// Unstakes some value in the specified era.
+    /// User should ensure that given era is either equal or greater than the
+    /// latest available era in the staking info.
+    fn unstake(&mut self, current_era: EraIndex, value: Balance) -> Result<(), &str> {
+        if self.stakes.is_empty() {
+            return Ok(());
+        }
+
+        let era_stake = self.stakes.last().unwrap(); // not empty so it exists
+        if era_stake.era > current_era {
+            return Err("Unexpected era".into());
+        }
+
+        let new_stake_value = era_stake.staked.saturating_sub(value);
+
+        if current_era == era_stake.era {
+            *self.stakes.last_mut().unwrap() = EraStake::new(new_stake_value, current_era)
+        } else {
+            self.stakes
+                .push(EraStake::new(new_stake_value, current_era))
+        }
+
+        self.clean_unstaked();
+        Ok(())
+    }
+
+    /// `Claims` the oldest era available for claiming.
+    /// In case valid era exists, returns (claim era, staked amount) tuple.
+    /// If no valid era exists, returns (0, 0) tuple.
+    fn claim(&mut self) -> (EraIndex, Balance) {
+        if self.is_empty() {
+            (0, Zero::zero())
+        } else {
+            let era_stake = self.stakes[0]; // not empty so it exists
+
+            if self.stakes.len() == 1 || self.stakes[1].era > era_stake.era + 1 {
```

Comment:
What if `self.stakes[1]` does not exist?

Reply:
That code will never execute if the first part of the clause is true.

If `self.is_empty()` is false and `self.stakes.len() == 1` is false, it means that `self.stakes.len()` is greater than 1.
So `self.stakes[1]' will always exist if that part of the code is executed.

Reply:
We have two possible branches:
* first one is when we only have a single element in the vec OR if the next era for which we have info isn't the consecutive era which we are claiming. E.g.:
<1, 1000>, <5, 2000> ==> <2, 1000>, <5, 2000>
*In this scenario, we just modify the **first** element by changing its era
* the second scenario can only be the case where the next element in vec is the consecutive era
* this means we can just delete the first element since that information isn't relevant anymore
<1, 1000>, <2, 2000> ==> <2, 2000>